pkg/kubelet/kubelet.go

```diff
@@ -349,12 +369,37 @@ func (self *podWorkers) Run(podFullName string, action func()) {
 	}()
 }
 
-func makeEnvironmentVariables(container *api.Container) []string {
+func (kl *Kubelet) makeEnvironmentVariables(container *api.Container) ([]string, error) {
 	var result []string
+	// Add service environment variables.  Note:  These are added to the docker.Config, but are
+	// not included in the checksum computed by dockertools.BuildDockerName(...).  That way,
+	// we can still determine whether an api.Container is already running by its hash.
+	// (We don't want to restart a container just because some service changed.)
+	//
+	// TODO: limit env vars to same namespace services, plus kubernetes{-ro}. (Issue #3293)
+	//
+	// Note that there is a race between Kubelet seeing the pod and kubelet seeing the service.
+	// To avoid this users can: (1) wait between starting a service and starting; or (2) detect
+	// missing service env var and exit and be restarted; or (3) use DNS instead of env vars
+	// and keep trying to resolve the DNS name of the service (recommended).
+	serviceEnv, err := kl.getServiceEnvVarMap()
+	if err != nil {
+		return result, err
+	}
 	for _, value := range container.Env {
+		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
+		// So, the master may set service env vars, or kubelet may.  In case both are doing
+		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
+		// env vars.
+		// TODO: remove this net line once all platforms use apiserver+Pods.
+		delete(serviceEnv, value.Name)
```

Comment:
I am fine with you doing all the services things.

One gotcha: if the services watching can be turned on independent of pod
watching the it needs to deduplicate env vars set in boundpods.  Basically
three modes have to work right: both boundPod and Service from etcd;
boundPod etcd but service api; or pods (not bound) and service from api.
On Jan 9, 2015 11:15 PM, "Paul Morie" <notifications@github.com> wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3376#discussion-diff-22757132>
> :
>
> >  	for _, value := range container.Env {
> > +		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
> > +		// So, the master may set service env vars, or kubelet may.  In case both are doing
> > +		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
> > +		// env vars.
> > +		// TODO: remove this net line once all platforms use apiserver+Pods.
> > +		delete(serviceEnv, value.Name)
>
> @erictune <https://github.com/erictune> in order to solve that leakage
> problem effectively we will have to diff the sets of services first, then
> generte the variables for the resulting set in the kubelet. Going to
> continue in my branch with your adapted code and move the whole shebang out
> of etcd and into the kubelet in my PR. That would divide the work up along
> resources, if i'm not mistaken. My PR will cover services, yours will cover
> pods. Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3376/files#r22757132>
> .
>